### PR TITLE
feat(crd): enforce KongPlugin and KongClusterPlugin's `plugin` field be immutable via CRD validation rules

### PR DESCRIPTION
**Pull request by pangruoran**
**What this PR does / why we need it**:

part of #5062 

**Which issue this PR fixes**:

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: pkg/apis/configuration/v1/kongplugin_types.go [hunk 1]
```diff
@@ -35,6 +35,7 @@ import (
 // +kubebuilder:printcolumn:name="Config",type=string,JSONPath=`.config`,description="Configuration of the plugin",priority=1
 // +kubebuilder:printcolumn:name="Programmed",type=string,JSONPath=`.status.conditions[?(@.type=="Programmed")].status`
 // +kubebuilder:validation:XValidation:rule="(has(self.config) || has(self.configFrom)) ? ((!has(self.config) && has(self.configFrom)) || (has(self.config) && !has(self.configFrom))) : true", message="Using both config and configFrom fields is not allowed."
+// +kubebuilder:validation:XValidation:rule="has(self.plugin)", message="Need to provide plugin"
```
Comment (pangruoran) [inline pkg/apis/configuration/v1/kongplugin_types.go:38]: Should I change this message to "The plugin field needs to be provided"?

**Comment by randmonkey** [comment]
This is not so necessary because the `plugin` field is already specified to be required here: 
https://github.com/Kong/kubernetes-ingress-controller/blob/27755d7e25e660e60f163a4fae34ede869aff02b/config/crd/bases/configuration.konghq.com_kongclusterplugins.yaml#L245-L246
https://github.com/Kong/kubernetes-ingress-controller/blob/27755d7e25e660e60f163a4fae34ede869aff02b/config/crd/bases/configuration.konghq.com_kongplugins.yaml#L240-L241

**Comment by pangruoran** [comment]
Yes, it is already required.   It doesn't seem to bring significant benefits.


Before

```
➜  ~ kubectl apply -f kp.yaml                                                                                                                                                                  
The KongPlugin "add-header-route" is invalid: plugin: Required value
```

After

```
➜  ~ kubectl apply -f kp.yaml
The KongPlugin "add-header-route" is invalid: 
* plugin: Required value
* <nil>: Invalid value: "null": some validation rules were not checked because the object was invalid; correct the existing errors to complete validation
```


I have a question about it. Can the `plugin` field be modified after it has been created?  @randmonkey

**Comment by pangruoran** [comment]
I believe that for our use case, the `plugin` field of `KongPlugins`/`KongClusterPlugin` should be immutable.
I will revise this PR.

**Comment by pangruoran** [comment]
After this change:

```
➜  ~ kubectl apply -f  kp.yaml
The KongPlugin "add-header-route" is invalid: <nil>: Invalid value: "object": The plugin field should be immutable
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -87,6 +87,8 @@ Adding a new version? You'll need three changes:
     [#5119](https://github.com/Kong/kubernetes-ingress-controller/pull/5119)
   - `KongConsumer` now enforces that at least one of `username` or `custom_id` is provided.
     [#5137](https://github.com/Kong/kubernetes-ingress-controller/pull/5137)
+  - `KongPlugin` and `KongClusterPlugin` now enforce `plugin` is immutable.
```
Comment (randmonkey) [inline CHANGELOG.md:90]:
```suggestion
  - `KongPlugin` and `KongClusterPlugin` now enforce `plugin` to be immutable.
```

File: config/crd/bases/configuration.konghq.com_kongclusterplugins.yaml [hunk 3]
```diff
@@ -250,6 +250,8 @@ spec:
           rule: '(has(self.config) || has(self.configFrom)) ? ((!has(self.config)
             && has(self.configFrom)) || (has(self.config) && !has(self.configFrom)))
             : true'
+        - message: The plugin field should be immutable
```
Comment (randmonkey) [inline config/crd/bases/configuration.konghq.com_kongclusterplugins.yaml:253]:
```suggestion
        - message: The plugin field is immutable
```

File: config/crd/bases/configuration.konghq.com_kongplugins.yaml [hunk 4]
```diff
@@ -245,6 +245,8 @@ spec:
           rule: '(has(self.config) || has(self.configFrom)) ? ((!has(self.config)
             && has(self.configFrom)) || (has(self.config) && !has(self.configFrom)))
             : true'
+        - message: The plugin field should be immutable
```
Comment (randmonkey) [inline config/crd/bases/configuration.konghq.com_kongplugins.yaml:248]:
```suggestion
        - message: The plugin field is immutable
```

**Review by randmonkey — COMMENTED** [review]
Is it possible to add test cases to test the validation of "plugin is immutable"?

**Comment by pangruoran** [comment]
Sure, let me add test case

File: internal/admission/validator.go [hunk 5]
```diff
@@ -281,9 +281,6 @@ func (validator KongHTTPValidator) ValidatePlugin(
 	ctx context.Context,
 	k8sPlugin kongv1.KongPlugin,
 ) (bool, string, error) {
-	if k8sPlugin.PluginName == "" {
```
Comment (randmonkey) [inline internal/admission/validator.go:284]: I have a doubt that we could not remove the validation in webhooks here. The CEL is enabled in k8s 1.23 and enabled by default in 1.25, so if we still need to support 1.22+, we could not remove the validation in webhooks. WDYT? @pmalek

Comment (pmalek) [reply internal/admission/validator.go:284]: According to https://docs.konghq.com/kubernetes-ingress-controller/latest/reference/version-compatibility/#general and a [long chat on slack](https://kongstrong.slack.com/archives/C011RQPHDC7/p1698757312230499) we're supporting 1.25+ with v3.0

Comment (pangruoran) [reply internal/admission/validator.go:284]: yes. we're supporting 1.25+ with v3.0

**Review by randmonkey — APPROVED** [review]

File: internal/admission/validator.go (hunk 5 above)
Comment (pmalek) [reply internal/admission/validator.go:284]: Given that we already have this field set as required I'm OK with this change 👍 https://github.com/Kong/kubernetes-ingress-controller/blob/314951c3279dcd38cb7de7e53f71969379a0340a/pkg/apis/configuration/v1/kongplugin_types.go#L67

**Review by pmalek — APPROVED** [review]

**Pull request merged**